### PR TITLE
Add Circuit Assembler Mode

### PR DESCRIPTION
**Pull request by LewisSaber**
Adds circuit assembler mode to CAL
The minimum length for this mode to work is 5, I dont want players to make short and ugly CALs
Currently Textures for Buttons are from precise assembler, will change them when someone makes any in #texture-dev
image

### CODE REVIEW

**Comment by chochem** [comment]
hmm, wasnt crib specifically removed for CAL? I guess people wont use this for LuV+ recipes due to 6x cost, but for lower circuits CA mode is just a easy mode CAL? (no imprints, allowed cribs, no downsides)
if you want to keep all the upsides, one solution could be to remove the perfect OCs.

**Comment by lc-1337** [comment]
> hmm, wasnt crib specifically removed for CAL? I guess people wont use this for LuV+ recipes due to 6x cost, but for lower circuits CA mode is just a easy mode CAL? (no imprints, allowed cribs, no downsides) if you want to keep all the upsides, one solution could be to remove the perfect OCs.

Its about ULV, LV and MV circuits? Thats not a big affect for LuV+. But anyway every multi have benefits (speed/energy discount) compare to single block one, so why this one shoudnt. I dont think that should nerf whole multi because three recipe...

I would like just remove these circuit recipes from mCA. I mean we should use CAL for circuits, not mCA. But not sure if that possible without "weird solutions".

**Comment by chochem** [comment]
nah that doesnt go to MV. that is upto zpm.

Also you misunderstand. this isnt a benefit over singleblock. these are benefits over the *already existing* multiblock.

**Comment by SMSum** [comment]
I’m gonna go on a limb and say this circuit assembler mode in CAL is not for circuits but to help eliminate another usage of PA. More than just circuits are made in the CA unfortunately.

**Comment by chochem** [comment]
thats the intent yea.

**Comment by SMSum** [comment]
So why does it matter if it uses cribs, people would already be using cribs on their circass PA. I don’t personally see anything in the circuit assembler that would do harm to have perfect OC either.

**Review by chochem — DISMISSED** [review]
before people derail this further let me just make this a clear request for changes:

in its current form this makes CA mode strictly better than current CAL for many recipes. That makes no sense from a gameplay or balance perspective. When there is a CAL recipe for an item, people should be incentivized to use it, not use the easier CA mode.

There can be various solutions to this, removing perfect OC for CA mode is just one such option. I am happy to hear other proposals.

**Comment by SMSum** [comment]
> before people derail this further let me just make this a clear request for changes:
> 
> in its current form this makes CA mode strictly better than current CAL for many recipes. That makes no sense from a gameplay or balance perspective. When there is a CAL recipe for an item, people should be incentivized to use it, not use the easier CA mode.
> 
> There can be various solutions to this, removing perfect OC for CA mode is just one such option. I am happy to hear other proposals.

CA has always been more expensive but ultimately can be faster if you PA it, yet people don't really play that "meta" do they?
Even with a perfect OC I don't see many people being deranged enough to spend 6x mat cost later game. 

Here is my proposition, keep currently proposed CA mode OC's and buff the CAL. Essentially just shift the entire playing field to the right a little bit. CAL CircAss gets perfect OC, CAL normal gets a speed boost to match and still incentivize usage > CircAss mode

The CAL needs a little bit of a pick me up. In its current state, you're unfortunately forced to spam an absurd amount. And before you go and try to make a counterargument that "Making machines faster wont fix spam", take a second to reflect on the subtick update. MEBF's can now do billions of parallels and the result was that MEBF spam has been almost entirely eliminated.

**Comment by srdr2k3** [comment]
> before people derail this further let me just make this a clear request for changes:
> 
> in its current form this makes CA mode strictly better than current CAL for many recipes. That makes no sense from a gameplay or balance perspective. When there is a CAL recipe for an item, people should be incentivized to use it, not use the easier CA mode.
> 
> There can be various solutions to this, removing perfect OC for CA mode is just one such option. I am happy to hear other proposals.

Would it be possible to block recipes with a CAL alternative being done in the CA mode instead?
Eg: It would not be possible to make a Integrated Processor in the CA mode because there is already a CAL recipe for it.

Then that way the CA mode would not be more overpowered than the CAL.
Though, then there is the issue of getting circuits for the first imprints.

**Comment by chochem** [comment]
that would work yea, the first you can get with singleblock, same as currently.

**Comment by lc-1337** [comment]
Remove/Nerf GT circuits recipes then. This multi is replace for PA, right? Multi with one regular energy hatch and without OC as replace for PA...

Also processor assembly/supercomputers have 25% speed discount in CAL. so is really problem up to zpm?

**Comment by SMSum** [comment]
> Remove/Nerf GT circuits recipes then. This multi is replace for PA, right? Multi with one regular energy hatch and without OC as replace for PA...
> 
> Also processor assembly/supercomputers have 25% speed discount in CAL. so is really problem up to zpm?

Yea, 1 hatch, no perfect OC would be awful and doesn't help resolve the problem of trying to find replacements for PA's. Unless you give it an absurd number of parallels I don't see how that would entice people to not just spam singleblocks.

**Comment by LewisSaber** [comment]
The perfect OC is made to make it somewhere closer to speed of PA without adding laser support
Even with current perfect OC it takes 2,5UXV CALs to have same speed as downtiered UMV PA making 
16384k AE  storage components(10s at UV each) 
Now about circuits, all circuits that are made in CA are more expencive(IV+ circuits) and are slower
For example(time to make 16 circuits, taken circuits that are first step of their line using recipe with asmd)
         CA    CAL
MV   80s    60s
EV     80s    60s
LUV   640s 160s
There is no way someone will prefer CA recipes over CAL recipes

**Comment by chochem** [comment]
> Also processor assembly/supercomputers have 25% speed discount in CAL. so is really problem up to zpm?

ah nice. yes thats a good thing, that should already be enough. :)
There never was any reason for it to have a discount before, so I did not realize it already is built in.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/GT_TileEntity_CircuitAssemblyLine.java [hunk 1]
```diff
@@ -448,10 +528,12 @@ public void getWailaNBTData(EntityPlayerMP player, TileEntity tile, NBTTagCompou
         super.getWailaNBTData(player, tile, tag, world, x, y, z);
         String imprintedWith = this.getTypeForDisplay();
         if (!imprintedWith.isEmpty()) tag.setString("ImprintedWith", imprintedWith);
+        tag.setInteger(RUNNING_MODE_KEY, mode);
     }
 
     @Override
     protected boolean supportsCraftingMEBuffer() {
-        return false;
+        return mode == 0 ? false : true;
```
Comment (BlueWeabo) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/GT_TileEntity_CircuitAssemblyLine.java]: can't this be done with just mode == 1? or mode != 0

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/GT_TileEntity_CircuitAssemblyLine.java [hunk 2]
```diff
@@ -425,6 +492,16 @@ public boolean supportsBatchMode() {
         return true;
     }
 
+    @Override
+    public boolean supportsInputSeparation() {
+        return mode == 0 ? false : true;
```
Comment (BlueWeabo) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/GT_TileEntity_CircuitAssemblyLine.java]: can't this be done with just mode == 1? or mode != 0

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/GT_TileEntity_CircuitAssemblyLine.java [hunk 3]
```diff
@@ -425,6 +492,16 @@ public boolean supportsBatchMode() {
...
+    }
+
+    @Override
+    public boolean isInputSeparationEnabled() {
+        return mode == 0 ? false : super.isInputSeparationEnabled();
```
Comment (BlueWeabo) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/GT_TileEntity_CircuitAssemblyLine.java]: wouldn't super.isSeperationEnabled() && mode !=0 or super.isSeparationEnabled() && mode == 1. be better? won't need to use ternary.

**Comment by Dream-Master** [comment]
GDCloud — Yesterday at 4:41 AM
the perfect OC seems fine to me, the circuit assembler mode will never be faster than the regular one because of the inbuilt (16/3)x speed boost and it takes 6x materials for circuits too, so I doubt its gonna cause problems

**Review by BlueWeabo — APPROVED** [review]
looks good code wise

**Pull request merged**